src/error.rs

```diff
@@ -22,8 +24,10 @@ pub enum Error {
     /// An `io::Error` that occurred while trying to read or write to a network stream.
     Io(IoError),
     /// An error from the `native_tls` library during the TLS handshake.
+    #[cfg(feature = "ssl")]
```

Comment:
This (and some of the other changes) would make the `ssl` feature non-additive, which [causes problems](https://github.com/Geal/nom/issues/544). I'm not entirely sure that there's a way to work around this :/

Reply:
Yes this is a tricky one to solve cleanly, especially without any breaking changes (split the enum into subset/superset relationship).
Perhaps it is appropriate though to allow this as the current API of the crate makes it a conscious decision to create a secure or insecure session - it would not be possible anyway to switch from not having the ssl feature to having it without also having to adapt other parts of a user's code.

Reply:
Unfortunately crate authors do not have the option of choosing. Since `cargo` assumes that features are additive, it will automatically pick the maximal set of features for any dependency. For example if `a` and `b` both depended on `imap`, one with the feature and one without, cargo would only compile `imap` once _with_ the feature and then compile both `a` and `b` against that, at which point `b` would fail to compile if it ever tried to match over the errors for example. That's far from ideal..